File: apps/web/src/lib/api/anki.ts
import { apiClient } from "./client"

export interface AnkiSyncResponse {
  synced: number
  queued: number
  pending_total: number
}

export interface AnkiSettingsResponse {
  anki_connect_url: string | null
  updated_at: string
}

export async function syncToAnki(languageId: number): Promise<AnkiSyncResponse> {
  return apiClient("/vocabulary/sync-anki", {
    method: "POST",
    body: JSON.stringify({ language_id: languageId }),
  })
}

export async function getAnkiStatus(languageId: number): Promise<AnkiSyncResponse> {
  return apiClient(`/vocabulary/anki-status?language_id=${languageId}`)
}

export async function getAnkiSettings(): Promise<AnkiSettingsResponse> {
  return apiClient("/admin/anki/settings")
}

export async function updateAnkiSettings(url: string | null): Promise<AnkiSettingsResponse> {
  return apiClient("/admin/anki/settings", {
    method: "PATCH",
    body: JSON.stringify({ anki_connect_url: url }),
  })
}

export async function testAnkiConnection(): Promise<{ success: boolean; message: string }> {
  return apiClient("/admin/anki/test", { method: "POST" })
}

File: apps/backend/src/api/routes/admin/anki.py
from datetime import datetime
from typing import Optional

import httpx
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.dependencies import get_db, require_admin
from src.infrastructure.db.models.users import User
from src.infrastructure.db.repositories.anki_repo import AnkiRepository

router = APIRouter(prefix="/admin/anki", tags=["admin"])
_repo = AnkiRepository()


class AnkiSettingsResponse(BaseModel):
    anki_connect_url: Optional[str]
    updated_at: datetime

    model_config = {"from_attributes": True}


class UpdateAnkiSettingsRequest(BaseModel):
    anki_connect_url: Optional[str] = None


class TestConnectionResponse(BaseModel):
    success: bool
    message: str


@router.get("/settings", response_model=AnkiSettingsResponse)
async def get_settings(
    _: User = Depends(require_admin),
    session: AsyncSession = Depends(get_db),
) -> AnkiSettingsResponse:
    settings = await _repo.get_settings(session)
    await session.commit()
    return AnkiSettingsResponse.model_validate(settings)


@router.patch("/settings", response_model=AnkiSettingsResponse)
async def update_settings(
    body: UpdateAnkiSettingsRequest,
    _: User = Depends(require_admin),
    session: AsyncSession = Depends(get_db),
) -> AnkiSettingsResponse:
    settings = await _repo.update_url(session, body.anki_connect_url)
    await session.commit()
    return AnkiSettingsResponse.model_validate(settings)


@router.post("/test", response_model=TestConnectionResponse)
async def test_connection(
    _: User = Depends(require_admin),
    session: AsyncSession = Depends(get_db),
) -> TestConnectionResponse:
    settings = await _repo.get_settings(session)
    if not settings.anki_connect_url:
        return TestConnectionResponse(
            success=False, message="No AnkiConnect URL configured"
        )

    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            resp = await client.post(
                settings.anki_connect_url,
                json={"action": "version", "version": 6},
            )
            resp.raise_for_status()
            data = resp.json()
            version = data.get("result", "unknown")
            return TestConnectionResponse(
                success=True, message=f"Connected — AnkiConnect v{version}"
            )
    except Exception as exc:
        return TestConnectionResponse(success=False, message=str(exc))
